### PR TITLE
Fix default initialization value in C API header

### PR DESCRIPTION
**Pull request by hariharans29**
**Description**: https://github.com/microsoft/onnxruntime/pull/9064/ initialized `do_copy_in_default_stream` in the `OrtCUDAProviderOptions` to 0. It needs to be initialized to 1 (which is the internal default as well). 

**Motivation and Context**
Please see https://github.com/microsoft/onnxruntime/issues/4829 for possible implications if it is set to 0. Also added a warning referencing the issue for users who want to set it to 0.

### CODE REVIEW

File: include/onnxruntime/core/session/onnxruntime_c_api.h [hunk 1]
```diff
@@ -345,29 +345,57 @@ typedef enum OrtCudnnConvAlgoSearch {
...
+  *   Defaults to 0.
+  *   \note If a ::OrtArenaCfg has been applied, it will override this field
   */
   int arena_extend_strategy;
+
+  /** \brief Flag indicating if copying needs to take place on the same stream as the compute stream in the CUDA EP   
+  *   0 = Use separate streams for copying and compute.
+  *   1 = Use the same stream for copying and compute.
+  *   Defaults to 1.
+  *   WARNING: Setting this to 0 may result in data races for some models.
```
Comment (hariharans29) [inline include/onnxruntime/core/session/onnxruntime_c_api.h:380]: Warning

File: include/onnxruntime/core/session/onnxruntime_c_api.h [hunk 2]
```diff
@@ -268,10 +268,10 @@ typedef OrtStatus* OrtStatusPtr;
 * When an allocator is passed to any function, be sure that the allocator object is not destroyed until the last allocated object using it is freed.
 */
 typedef struct OrtAllocator {
-  uint32_t version; ///< Must be initialized to ORT_API_VERSION
-  void*(ORT_API_CALL* Alloc)(struct OrtAllocator* this_, size_t size); ///< Returns a pointer to an allocated block of `size` bytes
-  void(ORT_API_CALL* Free)(struct OrtAllocator* this_, void* p); ///< Free a block of memory previously allocated with OrtAllocator::Alloc
-  const struct OrtMemoryInfo*(ORT_API_CALL* Info)(const struct OrtAllocator* this_); ///< Return a pointer to an ::OrtMemoryInfo that describes this allocator
+  uint32_t version;                                                                   ///< Must be initialized to ORT_API_VERSION
```
Comment (pranavsharma) [inline include/onnxruntime/core/session/onnxruntime_c_api.h:271]: Later we should enforce this.

**Review by pranavsharma — APPROVED** [review]

File: include/onnxruntime/core/session/onnxruntime_c_api.h (hunk 1 above)
Comment (pranavsharma) [inline include/onnxruntime/core/session/onnxruntime_c_api.h:380]: If we want this to be 1, ideally, we shouldn't even give the user an option to configure this. But we can't change the struct now. So, we should at least enforce this or else it'll lead to support issues. Can be done in a separate PR.

Comment (hariharans29) [reply include/onnxruntime/core/session/onnxruntime_c_api.h:380]: @KeDengMS added this field with the intention of giving the user an option.

File: include/onnxruntime/core/session/onnxruntime_c_api.h [hunk 3]
```diff
@@ -345,29 +345,57 @@ typedef enum OrtCudnnConvAlgoSearch {
 */
 typedef struct OrtCUDAProviderOptions {
 #ifdef __cplusplus
-  OrtCUDAProviderOptions() : device_id{}, cudnn_conv_algo_search{OrtCudnnConvAlgoSearchExhaustive}, gpu_mem_limit{SIZE_MAX}, arena_extend_strategy{}, do_copy_in_default_stream{}, has_user_compute_stream{}, user_compute_stream{}, default_memory_arena_cfg{} {}
+  OrtCUDAProviderOptions() : device_id{}, cudnn_conv_algo_search{OrtCudnnConvAlgoSearchExhaustive}, gpu_mem_limit{SIZE_MAX}, arena_extend_strategy{}, do_copy_in_default_stream{1}, has_user_compute_stream{}, user_compute_stream{}, default_memory_arena_cfg{} {}
 #endif
 
-  int device_id;  ///< CUDA device id (0 = default device)
+  /** \brief CUDA device Id
+  *   Defaults to 0.
```
Comment (edgchen1) [inline include/onnxruntime/core/session/onnxruntime_c_api.h:352]: identifying the default values is good. does it only apply to C++ usage though?

Comment (hariharans29) [reply include/onnxruntime/core/session/onnxruntime_c_api.h:352]: No, it is not limited to just C++.

Comment (edgchen1) [reply include/onnxruntime/core/session/onnxruntime_c_api.h:352]: was mainly wondering about the C API, as the constructor is within `#ifdef __cplusplus`

Comment (hariharans29) [reply include/onnxruntime/core/session/onnxruntime_c_api.h:352]: Can we have constructors for structs in C ? My understanding is for the users of the C API, the users are expected to explicitly initialize the struct's members.

Comment (edgchen1) [reply include/onnxruntime/core/session/onnxruntime_c_api.h:352]: that's my understanding too. would it be useful to clarify that the default value doesn't apply to C usage, as this is in the C API header?
or should we provide a way to get a structure with these default values that's also compatible with C?

Comment (hariharans29) [reply include/onnxruntime/core/session/onnxruntime_c_api.h:352]: I agree that there should be a way to fetch a default initialized instance for the convenience of the C API users. But can we maintain status quo for now and add the enhancement later ? Clarifying with a comment for this release is a nice middle ground. What do you think ?

**Review by wangyems — APPROVED** [review]

File: include/onnxruntime/core/session/onnxruntime_c_api.h (hunk 3 above)
Comment (edgchen1) [reply include/onnxruntime/core/session/onnxruntime_c_api.h:352]: sure, we can add it later

**Pull request merged**